Library/Homebrew/archive.rb

```diff
@@ -0,0 +1,185 @@
+# typed: false
+# frozen_string_literal: true
+
+require "digest/md5"
+require "utils/curl"
+
+# Archive API client.
+#
+# @api private
+class Archive
+  extend T::Sig
+
+  include Context
+  include Utils::Curl
+
+  class Error < RuntimeError
+  end
+
+  sig { returns(String) }
+  def inspect
+    "#<Archive: item=#{@archive_item}>"
+  end
+
+  sig { params(item: T.nilable(String)).void }
+  def initialize(item: "homebrew")
+    @archive_item = item
+
+    raise UsageError, "Must set the Archive item!" unless @archive_item
+
+    ENV["HOMEBREW_FORCE_HOMEBREW_ON_LINUX"] = "1" if @archive_item == "homebrew" && !OS.mac?
+  end
+
+  def open_api(url, *args, auth: true)
+    if auth
+      raise UsageError, "HOMEBREW_ARCHIVE_KEY is unset." unless (key = Homebrew::EnvConfig.archive_key)
+
+      if key.exclude?(":")
+        raise UsageError,
+              "Use HOMEBREW_ARCHIVE_KEY=access:secret. See https://archive.org/account/s3.php"
+      end
+
+      args += ["--header", "Authorization: AWS #{key}"]
+    end
+
+    curl(*args, url, print_stdout: false, secrets: key)
+  end
+
+  sig {
+    params(local_file:    String,
+           directory:     String,
+           remote_file:   String,
+           warn_on_error: T.nilable(T::Boolean)).void
+  }
+  def upload(local_file, directory:, remote_file:, warn_on_error: false)
+    unless File.exist? local_file
+      msg = "#{local_file} for upload doesn't exist!"
+      raise Error, msg unless warn_on_error
+
+      # Warn and return early here since we know this upload is going to fail.
+      opoo msg
+      return
+    end
+
+    md5_base64 = Digest::MD5.base64digest(File.read(local_file))
+    url = "https://#{@archive_item}.s3.us.archive.org/#{directory}/#{remote_file}"
```

Comment:
Would be good to make `s3.us.archive.org` a constant somewhere given it's used in a few places.